### PR TITLE
Form 

### PR DESCRIPTION
**Pull request by laxmikandel97**
Validation for form done and functionalities added. I need some reviews!

### CODE REVIEW

File: controller/Controller.php [hunk 1]
```diff
@@ -54,6 +53,7 @@ function tutorsPage($year)
      */
     function tutorsAjax()
     {
+        $this->_db->updateYearData($_POST["column"], $_POST["value"], $_POST["yearId"]);
```
Comment (kellerflint) [inline controller/Controller.php]: I don't think this line is supposed to be here. It should only happen inside the if statement (if "yearId" is set). I dont think changing it has anything to do with your page.

File: controller/Controller.php [hunk 2]
```diff
@@ -77,30 +78,95 @@ function checklistAjax(){
...
+            "W4" => array("Value" => $checkBoxes['year_w4'], "Column" => "year_w4")));
 
 
         $view = new Template();
         echo $view->render("views/checklist.html");
     }
+
+    /**
+     * Render page for form
+     * @param $param user's id
```
Comment (kellerflint) [inline controller/Controller.php]: @param datatype $variable is the format we're using for this project.

File: index.php [hunk 3]
```diff
@@ -27,15 +26,18 @@
 
 //create new Controller
 $controller = new Controller($f3, $db);
+$dirName = 'uploads/';
```
Comment (kellerflint) [inline index.php:30]: I would add a linebreak between this line and the Controller line and put a comment explaining that $dirName is the default place to upload files

File: model/database.php [hunk 4]
```diff
@@ -174,6 +175,92 @@ function testDatabase()
...
+    /**
+     * Insert into tutor table
+     * @param $firstName user's first name
+     * @param $lastName user's last name
+     * @param $phone user's phone
+     * @param $ssn user's ssn
+     * @return string last inserted id
+     * @author  laxmi
+     */
+    function insertMember($firstName, $lastName, $phone, $ssn)
```
Comment (kellerflint) [inline model/database.php]: You can probably delete this function, I don't think your form ever needs to insert a new member

File: model/database.php [hunk 5]
```diff
@@ -174,6 +175,92 @@ function testDatabase()
...
+        return $this->_dbh->lastInsertId();
+    }
+
+    /**
+     * update tutor table in given user's id
+     * @param $user_id  given user's id
+     * @param $firstName tutor's first name
+     * @param $lastName tutor's last name
+     * @param $phone tutor's phone
+     * @param $ssn tutor's ssn
```
Comment (kellerflint) [inline model/database.php]: @param datatype $variable for all of these

File: model/database.php [hunk 6]
```diff
@@ -174,6 +175,92 @@ function testDatabase()
...
+    function updateTutor($user_id, $firstName, $lastName, $phone, $ssn)
+    {
+        $sql = "UPDATE Tutor SET  tutor_first= ?, tutor_last=?,tutor_phone=?,tutor_ssn=? WHERE user_id=?";
+        $statement =$this->_dbh->prepare($sql);
+        $statement->execute([$firstName,$lastName,$phone,$ssn,$user_id]);
+    }
+
+    /**
+     * Get the tutor by their user_id
+     * @param $user_id given user's id
```
Comment (kellerflint) [inline model/database.php]: @param datatype $variable for all of these

File: model/database.php [hunk 7]
```diff
@@ -174,6 +175,92 @@ function testDatabase()
...
+        $sql = "SELECT * FROM Tutor WHERE user_id=?";
+        $statement =  $this->_dbh->prepare($sql);
+        $statement-> execute([$user_id]);
+        return $statement->fetch(PDO::FETCH_ASSOC);
+    }
+
+    /**
+     * Get user by it's id
+     * @param $user_id given user's id
+     * @return returns a rows from Tutor table
```
Comment (kellerflint) [inline model/database.php]: @param datatype $variable for all of these

File: model/database.php [hunk 8]
```diff
@@ -174,6 +175,92 @@ function testDatabase()
...
+        $sql = "SELECT * FROM User WHERE user_id =? ";
+        $statement = $this->_dbh-> prepare($sql);
+        $statement ->execute([$user_id]);
+        return $statement->fetch(PDO::FETCH_ASSOC);
+    }
+
+    /**
+     * Update email
+     * @param $user_id given user's id
+     * @param $email tutor's email
```
Comment (kellerflint) [inline model/database.php]: @param datatype $variable for all of these

File: model/database.php [hunk 9]
```diff
@@ -174,6 +175,92 @@ function testDatabase()
...
+    {
+        $sql = "UPDATE User SET user_email=? WHERE user_id =?";
+        $statement = $this->_dbh->prepare($sql);
+        $statement ->execute([$email,$user_id]);
+    }
+
+    /**
+     * Upload tutor's image
+     * @param $filePath path of image
+     * @param $user_id given user's id
```
Comment (kellerflint) [inline model/database.php]: @param datatype $variable for all of these

File: model/db_schema.sql [hunk 10]
```diff
@@ -65,16 +65,16 @@ INSERT INTO User Values (default,"Bobby@mail.com","bobTroll!!",0),(default,"Andy
...
+INSERT INTO Tutor VALUES ("1","Bob","Riely","(222) 222-4444","234-43-7853","image.jpg"),("2","Andy","Shmuck","253-786-5426","937-20-0573","image.jpg"),
+("3","Richard","Fanny","(222) 222-4444","073-96-1074","image.jpg"),("4","Jasmine","Riely","206-037-1056","658-43-0123","image.jpg"),
+("5","Reddin","Huck","(223) 232-4544","106-54-8528","image.jpg"),
+("7","Ben","Luss","(242) 222-4444","973-91-7893","image.jpg"),("8","Nick","Goravskiy","206-632-0835","123-45-6789","image.jpg"),
+("9","Joe","Carole","(252) 222-4444","964-53-0274","image.jpg"),("10","Ivan","Hambug","206-635-8468","936-02-6482","image.jpg"),
+("11","Gyser","Huck","(292) 022-4844","294-03-7851","image.jpg"),("12","Chad","Getter","253-635-9732","087-66-9882","image.jpg"),
+(13, "Some", "Guy", "(206) 222-4444",null,null),
+(14, "Tuna", "Ostrich", "(2062) 222-4644",null,null),
+(15, "Dave", "Test", "(202) 222-4444",null,null),
+(16, "Joe", "Smith", "(292) 222-4444",null,null);
```
Comment (kellerflint) [inline model/db_schema.sql:77]: Good thinking to update the test data. Much appreciated!

File: model/validate.php [hunk 11]
```diff
@@ -0,0 +1,189 @@
+<?php
+
+/**
+ * Validating onboarding form
+ * @author Laxmi kandel
+ */
+class Validate
+{
+    /**
+     * @var array
```
Comment (kellerflint) [inline model/validate.php]: this should not be a PHP Docs comment. It should be an inline comment with more description of the $_errors variable stores "Stores an array of error messages..." or whatever it does.

File: model/validate.php [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+     * @return array
+     */
+    public function getErrors()
+    {
+        return $this->_errors;
+    }
+
+    /**
+     * Validating First Name
+     * @param $firstName tutor's first name
```
Comment (kellerflint) [inline model/validate.php]: @param datatype $variable

File: model/validate.php [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+     */
+    function validFirstName($firstName)
+    {
+        $firstName = trim($firstName);
+        return ctype_alpha($firstName);
+    }
+
+    /**
+     * Validating Last Name
+     * @param $last tutor's last name
```
Comment (kellerflint) [inline model/validate.php]: @param datatype $variable

File: model/validate.php [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+     */
+    function validLastName($last)
+    {
+        $last = trim($last);
+        return ctype_alpha($last);
+    }
+
+    /**
+     * Validating email
+     * @param $email tutor's email
```
Comment (kellerflint) [inline model/validate.php]: @param datatype $variable

File: model/validate.php [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+        $emailResult = false;
+        if (filter_var($email, FILTER_VALIDATE_EMAIL)) {
+            $emailResult = true;
+        }
+        return $emailResult;
+    }
+
+    /**
+     * Validating Phone Number
+     * @param $phone tutor's phone
```
Comment (kellerflint) [inline model/validate.php]: @param datatype $variable

File: model/validate.php [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+        $phone = trim($phone);
+        if (strlen($phone) == 14 && preg_match($regex,$phone)){
+            $phoneResult = true;
+        }
+        return $phoneResult;
+    }
+
+    /**
+     * Validating SSN if provided
+     * @param ssn tutor's ssn
```
Comment (kellerflint) [inline model/validate.php]: @param datatype $variable

File: model/validate.php [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+            }
+        }else{
+            $ssnResult =true;
+        }
+        return $ssnResult;
+    }
+
+    /**Validating all the required fields name, phone, email, ssn, image
+     * @param $file user's selected file for image
+     * @param $newName name for file
```
Comment (kellerflint) [inline model/validate.php]: @param datatype $variable for all of these

File: model/validate.php [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+     */
+    public function getErrors()
+    {
+        return $this->_errors;
+    }
+
+    /**
+     * Validating First Name
+     * @param $firstName tutor's first name
+     * @return bool
```
Comment (kellerflint) [inline model/validate.php]: You use @return bool several times in this file and all of them need a description of what is being returned. Tell readers what it means when the function returns true

File: model/validate.php [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+                }
+            }
+        }
+        return $isValid;
+    }
+
+    /**
+     * Validate image
+     * @param $file image file
+     * @param $newName image name
```
Comment (kellerflint) [inline model/validate.php]: @param datatype $variable for all of these

File: styles/formStyleSheet.css [hunk 20]
```diff
@@ -33,4 +33,18 @@ legend {
...
+    width: 200px;
+}
+.err{
+    text-align: center;
+}
+em{
+    color: palevioletred;
+    font-size: smaller;
+    margin-left: 5px;
 }
```
Comment (kellerflint) [inline styles/formStyleSheet.css]: stylesheets should be in alphabetical order

File: views/form.html [hunk 21]
```diff
@@ -1,9 +1,9 @@
 <!--
-Form for OnBoarding process
-@author Laxmi Kandel
--->
+   Form for OnBoarding process
+   @author Laxmi Kandel
+   -->
 <!doctype html>
-<html lang="en">
+<html lang="en" xmlns="http://www.w3.org/1999/html">
```
Comment (kellerflint) [inline views/form.html]: I don't know what the xmlns="http://www.w3.org/1999/html" means but i'd guess you don't need it

File: views/form.html [hunk 22]
```diff
@@ -21,66 +21,125 @@ <h1 class="display-4">Onboarding Process form</h1>
...
         crossorigin="anonymous"></script>
 <script src="https://cdn.jsdelivr.net/npm/popper.js@1.16.0/dist/umd/popper.min.js"
         integrity="sha384-Q6E9RHvbIyZFJoft+2mJbHaEWldlvI9IOYy5n3zV9zzTtmI3UksdQRVvoxMfooAo"
         crossorigin="anonymous"></script>
 <script src="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/js/bootstrap.min.js"
         integrity="sha384-wfSDF2E50Y2D1uUdj0O3uMBJnjuUD4Ih7YwaYd1iqfktj0Uod8GCExl3Og8ifwB6"
         crossorigin="anonymous"></script>
+<script src="https://ajax.googleapis.com/ajax/libs/jquery/3.3.1/jquery.min.js"></script>
+<script src="https://unpkg.com/jquery-input-mask-phone-number@1.0.9/dist/jquery-input-mask-phone-number.js"></script>
+<script>
```
Comment (kellerflint) [inline views/form.html]: No inline JavaScript for production code. Please move this to a separate .js file

**Review by kellerflint — CHANGES_REQUESTED** [review]
Good overall, this is a lot of work! I have a few comments, mostly about changing how you write you PHP Docs so they are formatted the same way as everything else and a few other minor changes involving deleting unnecessary functions and moving js to a separate file.

**Comment by laxmikandel97** [comment]
Thank you Keller for review  I fixed all those!

**Comment by kellerflint** [comment]
You missed a couple of @param fixes and a thing in the controller. Check out the files changed tab, they should still be showing there.

**Review by kellerflint — APPROVED** [review]
Nice!

**Pull request merged**